experiments/MoarCSets/SInstances.jl

```diff
@@ -0,0 +1,406 @@
+  end
+  function SInstance{CD,AD,Ts,Idxed,TT,IT}() where
+    {CD <: CatDesc, AD <: AttrDesc{CD}, Ts <: Tuple,Idxed,TT <: NamedTuple,IT <: NamedTuple}
+    SInstance{CD,AD,Ts,Idxed}()
+  end
+  function SInstance{CD,AD,Ts,Idxed,TT,IT}(tables::TT,index::IT) where
+    {CD <: CatDesc, AD <: AttrDesc{CD}, Ts <: Tuple,Idxed,TT <: NamedTuple,IT <: NamedTuple}
+    new{CD,AD,Ts,Idxed,TT,IT}(tables,index)
+  end
+end
+
+
+function fields(::Type{T}) where {T <: NamedTuple}
+  T.parameters[1]
+end
+
+function fieldtype(::Type{T},i::Integer) where {T <: NamedTuple}
+  fieldtypes(T)[i]
+end
+
+function fieldtype(::Type{T},s::Symbol) where {T <: NamedTuple}
+  i = findfirst(fields(T) .== s)
+  fieldtype(T,i)
+end
+
+function fields(::Type{T}) where {T <: StructArray{<:NamedTuple}}
+  fields(T.parameters[1])
+end
+
+function Base.fieldtypes(::Type{T}) where {T <: StructArray{<:NamedTuple}}
+  fieldtypes(T.parameters[1])
+end
+
+function fieldtype(::Type{T},i::Integer) where {T <: StructArray{<:NamedTuple}}
+  fieldtype(T.parameters[1],i)
+end
+
+function fieldtype(::Type{T},s::Symbol) where {T <: StructArray{<:NamedTuple}}
+  fieldtype(T.parameters[1],s)
+end
+
+function Base.:(==)(x1::T, x2::T) where T <: SInstance
+  x1.tables == x2.tables
+end
+
+function Base.copy(ins::T) where T <: SInstance
+  T(copy(ins.tables),copy(ins.index))
+end
+
+Base.empty(ins::T) where T <: SInstance = T()
+
+function Base.show(io::IO, mime::MIME"text/plain", ins::AbstractSInstance{CD,AD,Ts}) where {CD,AD,Ts}
+  println(io,"SInstance")
+  for ob in CD.ob
+    println(io, "  $ob = 1:$(nparts(ins,ob))")
+  end
+  for (i,hom) in enumerate(CD.hom)
+    println(io, "  $hom : $(CD.ob[CD.dom[i]]) → $(CD.ob[CD.codom[i]])")
+    println(io, "    $(subpart(ins,hom))")
+  end
+  for (i,conc) in enumerate(AD.data)
+    println(io, "  $conc = $(Ts.parameters[i])")
+  end
+  for (i,attr) in enumerate(AD.attr)
+    println(io, "  $attr : $(CD.ob[AD.adom[i]]) ⇒ $(AD.data[AD.acodom[i]])")
+    println(io, "    $(subpart(ins,attr))")
+  end
+end
+
+nparts(ins::SInstance,type::Symbol) = length(ins.tables[type])
+
+haspart(ins::SInstance,type::Symbol,part::Int) = 1 <= part <= nparts(ins,type)
+
+""" Insert into sorted vector, preserving the sorting.
+"""
+function insertsorted!(a::AbstractVector, x)
+  insert!(a, searchsortedfirst(a, x), x)
+end
+
+""" Delete one occurrence of value from sorted vector.
+"""
+function deletesorted!(a::AbstractVector, x)
+  i = searchsortedfirst(a, x)
+  @assert i <= length(a) && a[i] == x "Value $x is not contained in $a"
+  deleteat!(a, i)
+end
+
+add_part!(ins::SInstance,type::Symbol,subparts::NamedTuple) =
+  _add_parts!(ins,Val(type),StructArray([subparts]))[1]
+add_parts!(ins::SInstance,type::Symbol,subpartses::StructArray{<:NamedTuple}) =
+  _add_parts!(ins,Val(type),subpartses)
+
+@generated function _add_parts!(ins::SInstance{CD,AD,Ts,Idxed,TT},::Val{ob},
+                                subpartses::T) where
```

Comment:
You don't have to change it, but when I read `subpartses` and I have to read it [like this](https://lotr.fandom.com/wiki/Gollum).

Reply:
Yeah, that was intentional :P

Reply:
Tasty subpartses.....